File: solutions/3161. Block Placement Queries/3161.cpp
// Time complexity: O((|queries| + |blocks|) * log(50001))
// Space complexity: O(50001 + |queries| + |blocks|)
class Solution {
 public:
  vector<bool> getResults(const vector<vector<int>>& queries) {
    vector<bool> ans;
    set<int> obstacles{0};

    for (const auto &query : queries)
      if (query[0] == 1)
        obstacles.insert(query[1]);

    for (auto it = next(begin(obstacles)); it != end(obstacles); ++it)
      updateBlockSize(*it, *it - *prev(it));

    for (int i = queries.size() - 1; i >= 0; --i) {
      int type = queries[i][0];
      int pos = queries[i][1];
      int blockSize = type == 1 ? 0 : queries[i][2];
      auto obstacleIt = obstacles.lower_bound(pos);

      if (type == 1) {
        if (next(obstacleIt) != end(obstacles))
          updateBlockSize(*next(obstacleIt), *next(obstacleIt) - 
                                             *prev(obstacleIt));
        obstacles.erase(obstacleIt);
      } else {
        bool canPlace = pos - *prev(obstacleIt) >= blockSize || 
                        getMaxBlockSize(pos) >= blockSize;
        ans.push_back(canPlace);
      }
    }

    return vector<bool>(rbegin(ans), rend(ans));
  }

 private:
  int blockSizeBIT[50001] = {};

  int getMaxBlockSize(int right) {
    int maxBlockSize = 0;

    for (; right >= 0; right = (right & (right + 1)) - 1)
      maxBlockSize = max(maxBlockSize, blockSizeBIT[right]);

    return maxBlockSize;
  }

  void updateBlockSize(int index, int value) {
    for (; index < 50001; index = index | (index + 1))
      blockSizeBIT[index] = max(blockSizeBIT[index], value);
  }
};

File: solutions/3161. Block Placement Queries/3161.py
# Time complexity: O((|queries| + |blocks|) * log(50001))
# Space complexity: O(50001 + |queries| + |blocks|)
class Solution:
  def __init__(self):
    self.segmentTree = []
    self.ans = []

  def getResults(self, queries: List[List[int]]) -> List[bool]:
    self.segmentTree = [0] * (min(50000, 3 * len(queries)) + 1)
    blocks = [0] + sorted(x for t, x, *_ in queries if t == 1)
    self._updateBlocks(blocks)

    for queryType, x, *size in reversed(queries):
      pos = bisect_left(blocks, x)
      if queryType == 1:
        if pos + 1 < len(blocks):
          self._updateSegmentTree(blocks[pos + 1], 
                                  blocks[pos + 1] - blocks[pos - 1])
        del blocks[pos]
      else:
        canPlace = x - blocks[pos - 1] >= size[0] or \
                    self._querySegmentTree(x) >= size[0]
        self.ans.append(canPlace)

    return list(reversed(self.ans))

  def _updateSegmentTree(self, pos: int, value: int) -> None:
    while pos < len(self.segmentTree):
      self.segmentTree[pos] = max(self.segmentTree[pos], value)
      pos |= (pos + 1)

  def _querySegmentTree(self, pos: int) -> int:
    maxValue = 0
    while pos >= 0:
      maxValue = max(maxValue, self.segmentTree[pos])
      pos = (pos & (pos + 1)) - 1

    return maxValue

  def _updateBlocks(self, blocks: List[int]) -> None:
    for i, blockEnd in enumerate(blocks[1:], start=1):
      self._updateSegmentTree(blockEnd, blockEnd - blocks[i - 1])